loadtest/control/simulcontroller/actions.go

```diff
@@ -1831,3 +1832,152 @@ func (c *SimulController) createPostReminder(u user.User) control.UserActionResp
+// rather than finding a post with a permalink in the current channel.
+// That would require some parsing to find out a permalink.
+func (c *SimulController) clickPermalink(u user.User) control.UserActionResponse {
+	currentCh, err := u.Store().CurrentChannel()
+	if err != nil {
+		return control.UserActionResponse{Err: control.NewUserError(err)}
+	}
+
+	posts, err := u.Store().ChannelPosts(currentCh.Id)
+	if err != nil {
+		return control.UserActionResponse{Err: control.NewUserError(err)}
+	}
+	var postID string
+	for _, p := range posts {
+		// Found a permalink.
+		if index := strings.Index(p.Message, "/pl/"); index != -1 {
+			// If there are multiple permalinks found, we will click on the last one
+			// in the channel. This naturally leads to an effect of having clicked all
+			// permalinks which emulates organic behavior.
+			postID = p.Message[index+4 : index+30]
+		}
+	}
+	if postID == "" {
+		return control.UserActionResponse{Info: "no permalink found"}
+	}
+
+	// We fetch the post thread.
+	postIds, _, err := u.GetPostThreadWithOpts(postID, "", model.GetPostsOptions{
+		CollapsedThreads: true,
+		Direction:        "down",
+		PerPage:          25,
+	})
+	if err != nil {
+		return control.UserActionResponse{Err: control.NewUserError(err)}
+	}
+
+	oldestPost, err := u.Store().Post(postIds[0])
+	if err != nil && !errors.Is(err, memstore.ErrPostNotFound) {
+		return control.UserActionResponse{Err: control.NewUserError(err)}
+	}
+
+	ch, err := u.Store().Channel(oldestPost.ChannelId)
+	if err != nil {
+		return control.UserActionResponse{Err: control.NewUserError(err)}
+	}
+	// We fetch the channel if it's not present in state.
+	// This is similar to what the webapp does.
+	if ch == nil {
+		if err := u.GetChannel(oldestPost.ChannelId); err != nil {
+			return control.UserActionResponse{Err: control.NewUserError(err)}
+		}
+		// We fetch the channel data from store again.
+		ch, err = u.Store().Channel(oldestPost.ChannelId)
+		if err != nil {
+			return control.UserActionResponse{Err: control.NewUserError(err)}
+		}
+	}
+
+	// We check if membership data is already there. If not,
+	// we call getChannelMember, and add the user to the channel
+	// if it's not a private channel.
+	cm, err := u.Store().ChannelMember(ch.Id, u.Store().Id())
+	if err != nil {
+		return control.UserActionResponse{Err: control.NewUserError(err)}
+	}
+	if cm.UserId == "" {
+		err := u.GetChannelMember(ch.Id, u.Store().Id())
+		if err != nil {
+			// Resorting to string matching, because the API doesn't return the
+			// actual Response object.
+			if strings.Contains(err.Error(), "You do not have the appropriate permissions") {
+				if ch.Type != model.ChannelTypePrivate {
+					if err := u.AddChannelMember(ch.Id, u.Store().Id()); err != nil {
+						return control.UserActionResponse{Err: control.NewUserError(err)}
+					}
+				}
+			} else {
+				return control.UserActionResponse{Err: control.NewUserError(err)}
+			}
+		}
+	}
```

Comment:
This is a bit hard to follow. Thoughts on moving it into a separate helper function so we can return early and avoid the five-level indentation?

Reply:
I removed one layer. It looks okay to me now. Let me know if you still think a separate helper function is necessary.

Reply:
Ah, that's better, thanks!